### PR TITLE
Polyfill the new EventListener API and use `passive: true` where makes sense.

### PR DESCRIPTION
**Pull request by aghassemi**
With this polyfill we can pass an `options` object  as the third argument to `addEventListener` and `removeEventListener` regardless of browser support. The real need for this polyfill arises from the fact that browsers that expect third-argument to be the `boolean capture` , would interpret an object as truthy and set `capture` to `true`, which is not what we want.

Please note that actual individual options are **not** polyfilled (something like `passive` can't truly be polyfilled anyway)

Closes https://github.com/ampproject/amphtml/issues/4819 

See https://github.com/ampproject/amphtml/issues/5105 for Closure Compiler extern issues related to this change and hence the `@suppress` blocks.

/to @jridgewell for review
/cc @dvoytenko @cramforce

### CODE REVIEW

**Comment by jridgewell** [comment]
Code LGTM. I'd rather see this as an exported function, instead of a polyfill.

**Comment by aghassemi** [comment]
I don't feel strongly either way but considering FF and Chrome have it and Safari has it in nightly already, I think polyfill makes sense.

**Comment by cramforce** [comment]
One benefit of a function would be that we could make passive the default :)

**Comment by dvoytenko** [comment]
Let's think about this in terms of API: would we rather continue using `listen` and `listenOnce` types of APIs with our `unsubscribe` concept or do we want to be closer the native DOM `addEventListener`? I'm gravitating toward our own `listen` because I'd like to get closer to the auto-unsubscribe system in many cases when resources are freed. This is currently one of our biggest issues on the roadmap to PWA and A4A is to ensure that we free up resources on element/document deletion.

File: extensions/amp-carousel/0.1/scrollable-carousel.js [hunk 1]
```diff
@@ -267,8 +267,20 @@ export class AmpScrollableCarousel extends BaseCarousel {
...
+    // We don't preventDefault, tell the browser to optimize scrolling.
+    const options = {
+      passive: true,
+    };
+
+    /**
+     * TODO(aghassemi, #5105): Type check
+     * @suppress {checkTypes}
+     */
+    (() => {
```
Comment (dvoytenko) [inline extensions/amp-carousel/0.1/scrollable-carousel.js]: Why extra closure?

File: extensions/amp-carousel/0.1/scrollable-carousel.js [hunk 2]
```diff
@@ -267,8 +267,20 @@ export class AmpScrollableCarousel extends BaseCarousel {
...
-    });
+
+    // We don't preventDefault, tell the browser to optimize scrolling.
+    const options = {
+      passive: true,
+    };
+
+    /**
+     * TODO(aghassemi, #5105): Type check
+     * @suppress {checkTypes}
```
Comment (dvoytenko) [inline extensions/amp-carousel/0.1/scrollable-carousel.js]: What's complication with type checking? Is it still not in externs? If so, please file a bug against closure.

File: src/polyfills/event-listener.js [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+      return false;
+    },
+  });
+  return supportsOptions;
+}
+
+/**
+ * Decides whether capture argument should be true/false based on the value of
+ * then given options argument.
+ * @param {*} options
```
Comment (dvoytenko) [inline src/polyfills/event-listener.js]: `@return`

File: src/polyfills/event-listener.js [hunk 4]
```diff
@@ -0,0 +1,103 @@
...
+    return !!(options.capture);
+  } else {
+    return options;
+  }
+}
+
+/**
+ * @param {!Window} win
+ */
+export function install(win) {
```
Comment (dvoytenko) [inline src/polyfills/event-listener.js]: This is another thing that needs to be polyfiled in a friendly frame. If we do proceed with this, please also call this polyfill from `installPolyfillsInChildWindow` in `extensions-impl.js`.

File: extensions/amp-carousel/0.1/scrollable-carousel.js (hunk 2 above)
Comment (aghassemi) [inline extensions/amp-carousel/0.1/scrollable-carousel.js]: I talked to them, there is an open PR to update the extern, they said they will review and merge in a week or so.

File: extensions/amp-carousel/0.1/scrollable-carousel.js (hunk 1 above)
Comment (aghassemi) [inline extensions/amp-carousel/0.1/scrollable-carousel.js]: this way I can get @suppress to only apply to that code (I tried to do this with code blocks in {} too but looks like these annotation only work at function scope)

**Comment by aghassemi** [comment]
looks like consensus is not polyfilling, I will move this to `listen` and related methods. There are not many places that use `listen` et al., so I can manually check them to ensure we can do `passive: true` as default.

I will also create an issue for us to migrate all `addEventListeners` to use `listen` instead and then we should not allow it anymore in presubmit. There are a lot of them and for each we should make sure the default `passive: true` is not a breaking change so that change won't be part of this CL and something to do in chunks later.

sounds good?

**Comment by jridgewell** [comment]
👍

**Pull request closed without merging**